domparsing/xmlserializer-polyfill.js

```diff
@@ -0,0 +1,844 @@
+      // 3.Replace any occurrences of "&" in markup by "&amp;".
+      markup = markup.replace("&", "&amp;");
+      // 4.Replace any occurrences of "<" in markup by "&lt;".
+      markup = markup.replace("<", "&lt;");
+      // 5.Replace any occurrences of ">" in markup by "&gt;".
+      markup = markup.replace(">", "&gt;");
+      // 6.Return the value of markup.
+      return markup;
+   }
+
+   function serializeDocumentFragment(node, contextNamespace, namespacePrefixMap, generatedNamespacePrefixRef, requireWellFormed) {
+      // 1.Let markup the empty string.
+      var markup = "";
+      // 2.For each child child of node, in tree order, run the XML serialization algorithm on the child
+      //   given a context namespace namespace, a namespace prefix map prefix map, a reference to a generated
+      //   namespace prefix index prefix index, and flag require well-formed. Concatenate the result to markup.
+      [].slice.call(node.childNodes).forEach(function serializeChildNodesOfDocumentFragment(childNode) {
+         markup += xmlSerializationAlgorithm(childNode, contextNamespace, namespacePrefixMap, generatedNamespacePrefixRef, requireWellFormed);
+      });
+      // 3.Return the value of markup.
+      return markup;
+   }
+
+   function serializeProcessingInstruction(node, requireWellFormed) {
+      // 1.If the require well-formed flag is set (its value is true), and node's target contains a ":"
+      //   (U+003A COLON) character or is an ASCII case-insensitive match for the string "xml", then throw
+      //   an exception; the serialization of this node's target would not be well-formed.
+      if (requireWellFormed && ((node.target.indexOf(":") >= 0) || (node.target.toLowerCase() == "xml")))
+         throw new Error("ProcessingInstruction's target contained invalid characters!");
+      // 2.If the require well-formed flag is set (its value is true), and node's data contains characters
+      //   that are not matched by the XML Char production [XML10] or contains the string "?>"
+      //   (U+003F QUESTION MARK, U+003E GREATER-THAN SIGN), then throw an exception; the serialization
+      //   of this node's data would not be well-formed.
+      if (requireWellFormed && (!matchXMLCharProduction(node.data) || (node.data.indexOf("?>") >= 0)))
+         throw new Error("ProcessingInstruction's data contained invalid characters");
+      // 3.Let markup be the concatenation of the following, in the order listed:
+      //   1."<?" (U+003C LESS-THAN SIGN, U+003F QUESTION MARK);
+      var markup = "<?";
+      //   2.The value of node's target;
+      markup += node.target;
+      //   3." " (U+0020 SPACE);
+      markup += " ";
+      //   4.The value of node's data;
+      markup += node.data;
+      //   5."?>" (U+003F QUESTION MARK, U+003E GREATER-THAN SIGN).
+      markup += "?>";
+      // 4.Return the value of markup.
+      return markup;
+   }
+
+   function matchXMLNameProduction(name) {
+      // TODO.
```

Comment:
This suggests that we don't have test coverage for this (and the similar checks below).